cumulusci/utils/clean_invalid_references.py

```diff
@@ -0,0 +1,365 @@
+
+    def return_name(self, element: ET.Element):
+        return element.find(self.name_xpath).text
+
+
+OBJECT_PERMISSION = PermissionElementXPath(".//objectPermissions", "object")
+RECORDTYPE_PERMISSION = PermissionElementXPath(
+    ".//recordTypeVisibilities", "recordType"
+)
+FIELD_PERMISSION = PermissionElementXPath(".//fieldPermissions", "field")
+USER_PERMISSION = PermissionElementXPath(".//userPermissions", "name")
+TAB_PERMISSION = PermissionElementXPath(".//tabVisibilities", "tab")
+APP_PERMISSION = PermissionElementXPath(".//applicationVisibilities", "application")
+APEXCLASS_PERMISSION = PermissionElementXPath(".//classAccesses", "apexClass")
+APEXPAGE_PERMISSION = PermissionElementXPath(".//pageAccesses", "apexPage")
+FLOW_PERMISSION = PermissionElementXPath(".//flowAccesses", "flow")
+CUSTOMPERM_PERMISSION = PermissionElementXPath(".//customPermissions", "name")
+CUSTOMSETTING_PERMISSION = PermissionElementXPath(".//customSettingAccesses", "name")
+CUSTOMMETADATA_PERMISSION = PermissionElementXPath(
+    ".//customMetadataTypeAccesses", "name"
+)
+
+PACKAGEXML_DICT = {
+    # PERMISSION : is_child
+    "CustomObject": {
+        OBJECT_PERMISSION: False,
+        CUSTOMSETTING_PERMISSION: False,
+        CUSTOMMETADATA_PERMISSION: False,
+        RECORDTYPE_PERMISSION: True,
+        FIELD_PERMISSION: True,
+    },
+    "ApexClass": APEXCLASS_PERMISSION,
+    "ApexPage": APEXPAGE_PERMISSION,
+    "Flow": FLOW_PERMISSION,
+    "CustomTab": TAB_PERMISSION,
+    "CustomApplication": APP_PERMISSION,
+    "CustomPermission": CUSTOMPERM_PERMISSION,
+}
+
+FOLDER_PERM_DICT = {
+    "objects": [OBJECT_PERMISSION, CUSTOMSETTING_PERMISSION, CUSTOMMETADATA_PERMISSION],
+    "fields": [FIELD_PERMISSION],
+    "tabs": [TAB_PERMISSION],
+    "applications": [APP_PERMISSION],
+    "customPermissions": [CUSTOMPERM_PERMISSION],
+    "pages": [APEXPAGE_PERMISSION],
+    "classes": [APEXCLASS_PERMISSION],
+    "flows": [FLOW_PERMISSION],
+    "recordTypes": [RECORDTYPE_PERMISSION],
+    "userPermissions": [USER_PERMISSION],
+}
+
+
+def run_queries_in_parallel(
+    queries: Dict[str, str], run_query: Callable[[str], dict], num_threads: int = 4
+):
+    """Accepts a set of queries structured as {'query_name': 'query'}
+    and a run_query function that runs a particular query. Runs queries in parallel and returns the queries"""
+    results_dict = {}
+
+    with ThreadPoolExecutor(max_workers=num_threads) as executor:
+        futures = {
+            query_name: executor.submit(run_query, query)
+            for query_name, query in queries.items()
+        }
+
+    for query_name, future in futures.items():
+        try:
+            query_result = future.result()
+            results_dict[query_name] = query_result
+        except Exception as e:
+            raise Exception(f"Error executing query '{query_name}': {type(e)}: {e}")
+        else:
+            queries.pop(query_name, None)
```

Comment:
I'm not sure I understand what this line does, since the loop doesn't iterate over `queries`. Is the goal to mutate the data structure for the caller?

Reply:
Ohh yes. That line is useless. I had earlier made this so that the same queries variable was reusable but now it doesnt matter. I will remove it